sql: crash on bad descriptor keys

I've randomly gotten into a crash which seems to be caused by some database descriptor keys not looking like the code expects - it wants an int and apparently it found a tuple.

I ran into this when starting crdb for a first time on a different machine than my usual one, and so I think the data dir was from an older version. Indeed, the log has this line:
```
I191229 03:08:54.666339 614 sql/event_log.go:132  [n1,intExec=set-version] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:19.2-6 User:root}
```

This is the panic:
```
E191229 03:10:12.578038 2298 sql/conn_executor.go:729  [n1,client=127.0.0.1:49832,user=root] a SQL panic has occurred while executing "SELECT t.oid, CASE WHEN nsp.nspname IN ('pg_catalog', 'public') THEN t.typname ELSE (nsp.nspname || '.') || t.typname END FROM pg_type AS t LEFT JOIN pg_type AS base_type ON t.typelem = base_type.oid LEFT JOIN pg_class AS base_cls ON base_type.typrelid = base_cls.oid LEFT JOIN pg_namespace AS nsp ON t.typnamespace = nsp.oid WHERE ((t.typtype IN ('b', 'p', 'r', 'e')) AND ((base_type.oid IS NULL) OR (base_type.typtype IN ('b', 'p', 'r'))))": value type is not INT: TUPLE
E191229 03:10:12.578438 2298 util/log/crash_reporting.go:207  [n1,client=127.0.0.1:49832,user=root] a panic has occurred!
panic: value type is not INT: TUPLE [recovered]
	panic: panic while executing 1 statements: SELECT _._, CASE WHEN _._ IN (_, _) THEN _._ ELSE (_._ || _) || _._ END FROM _ AS _ LEFT JOIN _ AS _ ON _._ = _._ LEFT JOIN _ AS _ ON _._ = _._ LEFT JOIN _ AS _ ON _._ = _._ WHERE ((_._ IN (_, _, __more2__)) AND ((_._ IS _) OR (_._ IN (_, _, __more1__)))); caused by value type is not INT: TUPLE
```

I'm thinking this has something to do with the migration to the new namespace table? Although the migrations appear to have finished before the crash, as indicated by this line:
```
I191229 03:08:54.651290 41 server/server.go:1635  [n1] done ensuring all necessary migrations have run
```

Full log: https://gist.github.com/andreimatei/8c8805021faeece8e387be21cb2c4b5b

What I've done is `cockroach workload init kv` and then `cockroach workload run kv`. The latter crashed. Perhaps the newly created database from the `init` have something to do with it.

Doesn't repro when restarting the server - but then again the data has been migrated and the databases have been created. Haven't tried harder.

Any idea what's going on? Perhaps it's obvious to @arulajmani?

--- a/pkg/sql/sqlbase/system.go
+++ b/pkg/sql/sqlbase/system.go
@@ -359,7 +359,10 @@ var (
 		NextMutationID: 1,
 	}
 
-	// NamespaceTable is the descriptor for the namespace table.
+	// NamespaceTable is the descriptor for the namespace table. Note that this
+	// table should only be written to via KV puts, not via the SQL layer. Some
+	// code assumes that it only has KV entries for column family 4, not the
+	// "sentinel" column family 0 which would be written by SQL.
 	NamespaceTable = TableDescriptor{
 		Name:                    "namespace",
 		ID:                      keys.NamespaceTableID,

--- a/pkg/sqlmigrations/migrations.go
+++ b/pkg/sqlmigrations/migrations.go
@@ -689,61 +689,56 @@ func createNewSystemNamespaceDescriptor(ctx context.Context, r runner) error {
 	return err
 }
 
+// migrateSystemNamespace migrates entries from the deprecated system.namespace
+// table to the new one, which includes a parentSchemaID column. Each database
+// entry is copied to the new table along with a corresponding entry for the
+// 'public' schema. Each table entry is copied over with the public schema as
+// as its parentSchemaID.
+//
+// New database and table entries continue to be written to the deprecated
+// namespace table until VersionNamespaceTableWithSchemas is active. This means
+// that an additional migration will be necessary in 20.2 to catch any new
+// entries which may have been missed by this one. In the meantime, namespace
+// lookups fall back to the deprecated table if a name is not found in the new
+// one.
 func migrateSystemNamespace(ctx context.Context, r runner) error {
-	// In system.namespace_deprecated, all databases have
-	// parentID == RootNamespaceID. When migrating databases over, the
-	// parentSchemaID is set to RootNamespaceID as well.
-	migrateDatabaseQuery := fmt.Sprintf(`
-INSERT INTO [%d AS namespace]("parentID", "parentSchemaID", name, id)
-(SELECT "parentID", %d, name, id FROM [%d AS namespace_deprecated] WHERE "parentID" = %d) 
-ON CONFLICT DO NOTHING`,
-		sqlbase.NamespaceTable.ID, keys.RootNamespaceID, sqlbase.DeprecatedNamespaceTable.ID, keys.RootNamespaceID,
-	)
-	// An entry for the `public` schema is added for every database in the old
-	// system.namespace table. This entry maps to a static ID of 29.
-	addPublicSchemasQuery := fmt.Sprintf(`
-INSERT INTO [%d AS namespace]("parentID", "parentSchemaID", name, id)
-(SELECT id, %d, 'public', %d FROM [%d AS namespace_deprecated] WHERE "parentID" = %d)
-ON CONFLICT DO NOTHING`,
-		sqlbase.NamespaceTable.ID, keys.RootNamespaceID, keys.PublicSchemaID, sqlbase.DeprecatedNamespaceTable.ID, keys.RootNamespaceID)
-	// In system.namespace_deprecated, all tables are implicitly under the
-	// `public` schema. When migrating tables over, the parentSchemaID is set
-	// to 29.
-	migrateTablesQuery := fmt.Sprintf(`
-INSERT INTO [%d AS namespace]("parentID", "parentSchemaID", name, id) 
-(SELECT "parentID", %d, name, id FROM [%d AS namespace_deprecated] WHERE "parentID" != %d)
-ON CONFLICT DO NOTHING`,
-		sqlbase.NamespaceTable.ID, keys.PublicSchemaID, sqlbase.DeprecatedNamespaceTable.ID, keys.RootNamespaceID,
-	)
-	_, err := r.sqlExecutor.ExecWithUser(
-		ctx,
-		"migrate-system-namespace-databases",
-		nil, /* txn */
-		security.NodeUser,
-		migrateDatabaseQuery,
-	)
+	q := fmt.Sprintf(
+		`SELECT "parentID", name, id FROM [%d AS namespace_deprecated]`,
+		sqlbase.DeprecatedNamespaceTable.ID)
+	rows, _, err := r.sqlExecutor.QueryWithUser(
+		ctx, "read-deprecated-namespace-table", nil /* txn */, security.NodeUser, q)
 	if err != nil {
 		return err
 	}
-	_, err = r.sqlExecutor.ExecWithUser(
-		ctx,
-		"migrate-system-namespace-add-public-schemas",
-		nil, /* txn */
-		security.NodeUser,
-		addPublicSchemasQuery,
-	)
-	if err != nil {
-		return err
-	}
-	_, err = r.sqlExecutor.ExecWithUser(
-		ctx,
-		"migrate-system-namespace-tables",
-		nil, /* txn */
-		security.NodeUser,
-		migrateTablesQuery,
-	)
-	if err != nil {
-		return err
+	for _, row := range rows {
+		parentID := sqlbase.ID(tree.MustBeDInt(row[0]))
+		name := string(tree.MustBeDString(row[1]))
+		id := sqlbase.ID(tree.MustBeDInt(row[2]))
+		if parentID == keys.RootNamespaceID {
+			// This row represents a database. Add it to the new namespace table.
+			databaseKey := sqlbase.NewDatabaseKey(name)
+			if err := r.db.Put(ctx, databaseKey.Key(), id); err != nil {
+				return err
+			}
+			// Also create a 'public' schema for this database.
+			schemaKey := sqlbase.NewSchemaKey(id, "public")
+			if err := r.db.Put(ctx, schemaKey.Key(), keys.PublicSchemaID); err != nil {
+				return err
+			}
+		} else {
+			// This row represents a table. Add it to the new namespace table with the
+			// schema set to 'public'.
+			if id == keys.DeprecatedNamespaceTableID {
+				// The namespace table itself was already handled in
+				// createNewSystemNamespaceDescriptor. Do not overwrite it with the
+				// deprecated ID.
+				continue
+			}
+			tableKey := sqlbase.NewTableKey(parentID, keys.PublicSchemaID, name)
+			if err := r.db.Put(ctx, tableKey.Key(), id); err != nil {
+				return err
+			}
+		}
 	}
 	return nil
 }

